packages/network-controller/src/create-network-client.ts

```diff
@@ -0,0 +1,217 @@
+  createRetryOnEmptyMiddleware,
+} from '@metamask/eth-json-rpc-middleware';
+import {
+  providerFromEngine,
+  providerFromMiddleware,
+  SafeEventEmitterProvider,
+} from '@metamask/eth-json-rpc-provider';
+import { createInfuraMiddleware } from '@metamask/eth-json-rpc-infura';
+import type { Hex } from '@metamask/utils';
+import { PollingBlockTracker } from 'eth-block-tracker';
+import { NetworksChainId, NetworkType } from '@metamask/controller-utils';
+
+const SECOND = 1000;
+
+export enum NetworkClientType {
+  Custom = 'custom',
+  Infura = 'infura',
+}
+
+type CustomNetworkConfiguration = {
+  chainId: Hex;
+  rpcUrl: string;
+  type: NetworkClientType.Custom;
+};
+
+export type InfuraNetworkType = Exclude<NetworkType, 'localhost' | 'rpc'>;
+
+type InfuraNetworkConfiguration = {
+  network: InfuraNetworkType;
+  infuraProjectId: string;
+  type: NetworkClientType.Infura;
+};
+
+/**
+ * Create a JSON RPC network client for a specific network.
+ *
+ * @param networkConfig - The network configuration.
+ * @returns the network client
+ */
+export function createNetworkClient(
+  networkConfig: CustomNetworkConfiguration | InfuraNetworkConfiguration,
+): { provider: SafeEventEmitterProvider; blockTracker: PollingBlockTracker } {
+  const rpcApiMiddleware =
+    networkConfig.type === NetworkClientType.Infura
+      ? createInfuraMiddleware({
+          network: networkConfig.network,
+          projectId: networkConfig.infuraProjectId,
+          maxAttempts: 5,
+          source: 'metamask',
+        })
+      : createFetchMiddleware({
+          btoa: global.btoa,
+          fetch: global.fetch,
+          rpcUrl: networkConfig.rpcUrl,
+        });
+
+  const rpcProvider = providerFromMiddleware(rpcApiMiddleware);
+
+  const blockTrackerOpts =
+    // eslint-disable-next-line node/no-process-env
+    process.env.IN_TEST && networkConfig.type === 'custom'
+      ? { pollingInterval: SECOND }
+      : {};
+  const blockTracker = new PollingBlockTracker({
+    ...blockTrackerOpts,
+    provider: rpcProvider,
+  });
+
+  const networkMiddleware =
+    networkConfig.type === NetworkClientType.Infura
+      ? createInfuraNetworkMiddleware({
+          blockTracker,
+          network: networkConfig.network,
+          rpcProvider,
+          rpcApiMiddleware,
+        })
+      : createCustomNetworkMiddleware({
+          blockTracker,
+          chainId: networkConfig.chainId,
+          rpcApiMiddleware,
+        });
+
+  const engine = new JsonRpcEngine();
+
+  engine.push(networkMiddleware);
+
+  const provider = providerFromEngine(engine);
+
+  return { provider, blockTracker };
+}
+
+/**
+ * Create middleware for infura.
+ *
+ * @param options0 - options for infura middleware
+ * @param options0.blockTracker - block tracker to use
+ * @param options0.network - network to use
+ * @param options0.rpcProvider - rpcProvider to use
+ * @param options0.rpcApiMiddleware - additional rpcApiMiddleware
+ * @returns the middleware
```

Comment:
Nit: Might be good to use `options` or `args` rather than the generated `options0` here. Also for consistency with our other JSDocs, would be good to capitalize the start of each description and end it with a period:

``` suggestion
 * @param args - The arguments.
 * @param args.blockTracker - The block tracker to use.
 * @param args.network - The Infura network to use.
 * @param args.rpcProvider - The RPC provider to use.
 * @param args.rpcApiMiddleware - Additional middleware.
 * @returns The middleware.
```